### PR TITLE
Hi-C Peak Calling with "StrawToBedPe"

### PR DESCRIPTION
**Pull request by rimangan**
# Description
This PR unveils the final vision for "StrawToBedPe", a program to call significant chromatin interactions in bedpe format from input Hi-C contact maps in Straw format.

image
image

Briefly, StrawToBedPe fits the distribution of contact scores for each distance to a zero-truncated negative binomial (ZTNB) distribution using ZTNB regression via coordinate ascent. Then, significant peaks are reported as significant outliers from these null contact distributions based on the total number of comparisons at each binDistance and a user-defined false discovery rate.

To this end, this PR also adds a "fit" numbers subpackage. This package houses functions for fitting statistical distributions to data. I have moved Dan's previous Gaussian mixture model code to this package, and added a new file that can fit Poisson or Negative Binomial distributions to data. 
To test the negative binomial fitting, I first generated random variates from negBinom distributions with known input parameters in R. 
image
I then used these as input data to the Gonomics curve fitting algorithm, and tests passed when our estimated parameters were within 5% of the true parameters.

I performed similar testing on the function that fits lambda distributions to input data.
image

## Relevant Links

- [Display Text](https://www.vertgenlab.org/)

### Checklist before requesting a review

- [ ] I performed a self-review of my code
- [ ] If it this a core feature, I have added thorough tests
- [ ] The command `go fmt` or `make clean` was used on all files included

### Testing

None

### Screenshots & Media

None

### Edge cases / Breaking Changes / Known Issues

None

### CODE REVIEW

**Comment by cfauci** [comment]
What's going on with all those testdata binary files?

**Comment by rimangan** [comment]
> What's going on with all those testdata binary files?

I made a bunch of NB and poisson distributed variates in R to test the vanilla NB/Poisson fit functions. These are just new line delimited text files of those numbers

File: cmd/strawToBedpe/strawToBedpe.go [hunk 1]
```diff
@@ -11,46 +11,278 @@ import (
...
 
+// makecomparisonCountCache is a helper function that calculates the number of comparisons made for each binDistance.
+func makeComparisonCountCache(contactScoreCache [][]int, searchSpaceMins map[string]int, searchSpaceMaxes map[string]int, s Settings) []int {
+	var totalWindows int
+	var currKey string
+	var comparisonCountCache []int = make([]int, len(contactScoreCache))
+	for i := range comparisonCountCache {
+		totalWindows = 0
+		for currKey, _ = range searchSpaceMins {
+			//fmt.Printf("currKey: %v. searchSpaceMins[currKey]: %v. searchSpaceMaxes[currKey]: %v. BinDistance: %v.\n", currKey, searchSpaceMins[currKey], searchSpaceMaxes[currKey], i)
```
Comment (cfauci) [inline cmd/strawToBedpe/strawToBedpe.go:191]: This looks like a debugging statement that can come out now

File: cmd/strawToBedpe/strawToBedpe.go [hunk 2]
```diff
@@ -11,46 +11,278 @@ import (
...
-	} else {
-		for s := range straw {
-			thisRec = bedpe.BedPe{A: bed.Bed{Chrom: chrom, ChromStart: s.Bin1Start, ChromEnd: s.Bin1Start + binSize, Score: s.ContactScore, FieldsInitialized: 8}, B: bed.Bed{Chrom: interChrom, ChromStart: s.Bin2Start, ChromEnd: s.Bin2Start + binSize, Score: s.ContactScore, FieldsInitialized: 8}}
-			bedpe.WriteToFileHandle(out, thisRec)
+	}
+
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
```
Comment (cfauci) [inline cmd/strawToBedpe/strawToBedpe.go:158]: go docs

**Review by cfauci — COMMENTED** [review]
I left some content comments about documentation and a debugging statement thats commented out but can be removed, but otherwise this looks good

**Review by cfauci — APPROVED** [review]

**Review by edotau — APPROVED** [review]
Was able to get through most of this, looks good from software stand point. Going to approve this to get it merge but if any issues arise we can open a follow up PR

Screenshot 2023-10-05 at 11 21 35 AM

**Pull request merged**